documentation/projects/proposals/trust_and_safety/content_report_moderation/20240122-implementation_plan_bulk_moderation_actions.md

```diff
@@ -0,0 +1,526 @@
+- By having a single model instead we get many things for free, such as:
+  - bulk decisions are visible in the Report view, which is valuable context for
+    moderators
+  - individual decisions are visible in the ModerationDecision view
+  - the DeletedMedia and SensitiveMedia views allow filtering by id for any type
+    of moderation decision
+
+For example, one option was to create a separate ModerationDecision for each
+record in a bulk decision, with an optional foreign_id to the
+BulkModerationDecision parent object. This would result in potentially massive
+duplication of information, however, if large bulk decisions are made.
+
+I also considered having `provider` and `creator` as columns on a special type
+of ModerationDecision for all of a particular creator's records.
+
+- This adds complexity and even more ModerationDecision types; alternatively,
+  these would be nullable columns that are not always applicable (again, adding
+  complexity)
+- May cause confusion by implying that a record which does have these fields
+  should apply to absolutely all records by that creator, when this may not be
+  the case (for example, if some records by this creator have already been
+  deindexed/marked sensitive by previous decisions)
+- Ultimately this context should be included in the `explanation` field
+  regardless
+
+### Having a special form for creating ModerationDecisions that affect all records by a creator
+
+I considered including the ability to add a ModerationDecision directly, with a
+modified form that would take `creator` and `provider` rather than selecting the
+individual records to add.
+
+However, this duplicates some work, and makes it much more confusing to explain
+the various ways that a decision can be made. It is also already possible to do
+this sort of bulk action by filtering in the Media views on creator/provider.
+
+### Having an "undo" action on individual decisions in the ModerationDecision list view
+
+This was an alternative to requiring the user to look up the id of the decision
+they want to reverse, and then filter for those records in the SensitiveMedia or
+DeletedMedia views to create a new decision.
+
+However this quickly becomes much more complicated. One example is that it does
+not make sense to undo a decision twice. It also makes it extremely difficult to
+add some useful features like:
+
+- Undoing a _subset_ of a decision, for example if we deindex all of a
+  particular creator's record but realize upon further investigation that only
+  some of them needed to be deindexed.
+- Undoing _all_ deindexing/marking sensitive for a particular creator, including
+  ones that were originally acted upon by individual reports.
+
+### A DAG for bulk decisions
+
+It's difficult to anticipate all the ways that we might need to filter records
+for selecting a set to take action on. Moreover, filtering in the Django admin
+views (especially by text search terms) may be extremely slow for some use
+cases.
```

Comment:
Can we can alleviate the slow search by using Elasticsearch instead? For creator searches I think it might actually be necessary (mentioned in another comment).

Reply:
Agreed in your earlier comment. I was fixated on changing the `search_fields`, but I think if we add all the new filters as custom filters we can do this in elasticsearch. 🚀